clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs

```diff
@@ -0,0 +1,387 @@
+{-|
+  Copyright   :  (C) 2021, QBayLogic B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Utilities to parse and render flags exposed on the Clash's command line.
+-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Clash.GHC.ClashFlag
+  ( ClashFlag(..)
+  , FlagName
+
+  -- * Simple flag builders
+  , boolFlag
+  , wordFlag
+  , intFlag
+  , readableFlag
+
+  , maybeStringFlag
+  , maybeIntFlag
+  , maybeWordFlag
+  , maybeReadableFlag
+
+  , maybeStringWithDefaultFlag
+
+  -- * Generic flag builders
+  , boolArg
+  , unsetArg
+  , wordSuffix, wordSuffixId
+  , intSuffix, intSuffixId
+  , strSuffix, strSuffixId
+
+  -- * Flag renderers
+  , renderNoArg
+  , renderSuffix
+  , renderSepSuffix
+
+  -- * Utilities
+  , deprecated
+  , isDefault
+  ) where
+
+-- base
+import Data.IORef (IORef, modifyIORef)
+
+-- ghc
+#if __GLASGOW_HASKELL__ >= 900
+import GHC.Driver.CmdLine
+#else
+import CmdLineParser
+#endif
+  (EwM, Flag (Flag, flagOptKind), liftEwM, OptKind (..), defFlag, addErr, deprecate)
+
+-- clash-lib
+import Clash.Driver.Types (ClashOpts(..), defClashOpts)
+
+-- lens
+import Control.Lens (Lens', (^.), set)
+import Text.Read (readMaybe)
+
+-----------
+-- Types --
+-----------
+
+-- | Defines a way to parse command line flags, set their values in a 'ClashOpts',
+-- and render flags based on a 'ClashOpts'.
```

Comment:
This is listed as three separate things, but I _think_ the parsing and setting are inherently related here. If that's the case it would be better to re-word and make it clear the two happen together / use bullet points